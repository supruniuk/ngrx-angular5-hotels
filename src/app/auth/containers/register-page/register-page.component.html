<mat-drawer-container>
  <div fxLayout="row" fxLayoutAlign="center center" class="register-container">
    <hot-register-form
      (onSubmit)="onSubmit($event)"
      (onCancel)="onCancel()"
      [pending]="pending$ | async"
      [errorMessage]="error$ | async">
    </hot-register-form>
  </div>
</mat-drawer-container>