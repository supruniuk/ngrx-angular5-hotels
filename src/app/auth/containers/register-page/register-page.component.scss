:host {
  height:100%;
  display:flex;
  flex-direction: column;
}

.register-container {
  height:100%;
}

mat-drawer-container {
  height: 100%;
}