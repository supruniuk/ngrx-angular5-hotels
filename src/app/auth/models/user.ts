export class Group {
  id: number;
  name: string;
}

export interface User {
  id: string;
  name: string;
  email: string;
  img: string;
  groups?: Group[];
}
