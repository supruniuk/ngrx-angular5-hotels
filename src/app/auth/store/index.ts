export * from './actions/auth';
export * from './reducers';
export * from './effects/auth.effects';
export * from './selectors';
