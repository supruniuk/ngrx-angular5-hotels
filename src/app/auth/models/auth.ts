import { User } from './user';


export interface Authenticate {
  email: string;
  password: string;
}

export interface SignUp extends Authenticate {
  authType: string;
  name: string;
}

export interface AuthResponse {
  user: User;
  token: string;
}

export const CUSTOMER_TYPE = 'customer';
export const HOTEL_TYPE = 'hotel';

export class Auth {
  constructor(
    public token: string,
    public type: string
  ) { }
}

// export class CustomerAuth extends Auth {
//   readonly type = CUSTOMER_TYPE;
// }

// export class HotelAuth extends Auth {
//   readonly type = HOTEL_TYPE;
// }
