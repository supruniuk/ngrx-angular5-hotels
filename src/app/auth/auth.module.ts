import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { ReactiveFormsModule } from '@angular/forms';
import { FlexLayoutModule } from '@angular/flex-layout';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import { Routes } from '@angular/router';

import { MaterialModule } from '../material';
import { LoginPageComponent } from './containers/login-page/login-page.component';
import { RegisterPageComponent } from './containers/register-page/register-page.component';
import { LoginFormComponent } from './components/login-form/login-form.component';
import { RegisterFormComponent } from './components/register-form/register-form.component';

import { AuthInterceptor, AuthService, AuthGuard, RoleGuard } from './services';
import { AuthEffects, reducers, metaReducers } from './store';


export const COMPONENTS = [
  LoginPageComponent,
  LoginFormComponent,
  RegisterPageComponent,
  RegisterFormComponent,
];

export const ROUTES: Routes = [
  { path: 'login', component: LoginPageComponent },
  { path: 'register', component: RegisterPageComponent }
];

@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MaterialModule,
    FlexLayoutModule
  ],
  declarations: COMPONENTS,
  exports: COMPONENTS,
})
export class AuthModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: RootAuthModule,
      providers: [
        AuthService,
        AuthGuard,
        RoleGuard,
        {
          provide: HTTP_INTERCEPTORS,
          useClass: AuthInterceptor,
          multi: true
        },
      ],
    };
  }
}

@NgModule({
  imports: [
    AuthModule,
    RouterModule.forChild(ROUTES),
    StoreModule.forFeature('auth', reducers, { metaReducers }),
    EffectsModule.forFeature([AuthEffects]),
  ],
})
export class RootAuthModule { }
