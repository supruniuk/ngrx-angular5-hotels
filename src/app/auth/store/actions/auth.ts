import { Action } from '@ngrx/store';
import { User } from '../../models/user';
import { Authenticate, SignUp, Auth } from '../../models/auth';

export enum AuthActionTypes {
  login = '[Auth] Login',
  logout = '[Auth] Logout',
  loginSuccess = '[Auth] Login Success',
  loginFail = '[Auth] Login Failure',
  loginRedirect = '[Auth] Login Redirect',
  customerRedirect = '[Auth] Customer Redirect',
  hotelRedirect = '[Auth] Hotel Redirect',
  register = '[Auth] Register',
  registerSuccess = '[Auth] Register Success',
  registerFail = '[Auth] Register Failure',
  registerRedirect = '[Auth] Register Redirect',
}

export class Login implements Action {
  readonly type = AuthActionTypes.login;

  constructor(public payload: Authenticate) {}
}

export class LoginSuccess implements Action {
  readonly type = AuthActionTypes.loginSuccess;

  constructor(public payload: { user: User, auth: Auth }) {}
}

export class LoginFail implements Action {
  readonly type = AuthActionTypes.loginFail;

  constructor(public payload: any) {}
}

export class LoginRedirect implements Action {
  readonly type = AuthActionTypes.loginRedirect;
}

export class CustomerRedirect implements Action {
  readonly type = AuthActionTypes.customerRedirect;
}

export class HotelRedirect implements Action {
  readonly type = AuthActionTypes.hotelRedirect;
}

export class Register implements Action {
  readonly type = AuthActionTypes.register;

  constructor(public payload: SignUp) {}
}

export class RegisterSuccess implements Action {
  readonly type = AuthActionTypes.registerSuccess;

  constructor(public payload: { user: User, auth: Auth }) {}
}

export class RegisterFail implements Action {
  readonly type = AuthActionTypes.registerFail;

  constructor(public payload: any) {}
}

export class RegisterRedirect implements Action {
  readonly type = AuthActionTypes.registerRedirect;

  constructor(public payload: string) {}
}

export class Logout implements Action {
  readonly type = AuthActionTypes.logout;
}

export type AuthActions = Login
  | LoginSuccess
  | LoginFail
  | LoginRedirect
  | CustomerRedirect
  | HotelRedirect
  | Register
  | RegisterSuccess
  | RegisterFail
  | RegisterRedirect
  | Logout;
