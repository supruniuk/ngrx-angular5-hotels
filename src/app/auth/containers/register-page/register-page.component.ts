import { Component, OnInit } from '@angular/core';
import { Store } from '@ngrx/store';
import { SignUp } from '../../models/auth';
import { AuthState } from '../../store/reducers';
import { getRegisterPagePending, getRegisterPageError } from '../../store/selectors';
import { Register, LoginRedirect } from '../../store/actions/auth';


@Component({
  selector: 'hot-register-page',
  templateUrl: './register-page.component.html',
  styleUrls: ['./register-page.component.scss'],
})
export class RegisterPageComponent implements OnInit {
  pending$ = this.store.select(getRegisterPagePending);
  error$ = this.store.select(getRegisterPageError);

  constructor(private store: Store<AuthState>) { }

  ngOnInit() { }

  onSubmit($event: SignUp) {
    this.store.dispatch(new Register($event));
  }

  onCancel() {
    this.store.dispatch(new LoginRedirect());
  }
}
