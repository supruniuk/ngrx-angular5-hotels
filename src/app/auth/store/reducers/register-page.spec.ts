import { reducer } from './register-page';
import * as fromRegisterPage from './register-page';
import { Register, RegisterSuccess, RegisterFailure, Logout } from '../actions/auth';
import { Authenticate, AuthResponse, User, SignUp } from '../models/user';

describe('RegisterPageReducer', () => {
  const user: User = {
    name: 'test_user',
    groups: [{ id: 1, name: 'test_group' }],
    email: 'test_email',
  };
  const authResp: AuthResponse = { user: user, token: 'test_token' };
  const authenticate: Authenticate = { email: 'test', password: 'password' };
  const signup: SignUp = {name: 'test_name', email: 'test@test.com', password: 'test_password' };

  describe('undefined action', () => {
    it('should return the default state', () => {
      const action = {} as any;

      const result = reducer(undefined, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('REGISTER', () => {
    it('should make pending to true', () => {
      const createAction = new Register(signup);

      const expectedResult = {
        error: null,
        pending: true,
      };

      const result = reducer(fromRegisterPage.initialState, createAction);

      expect(result).toMatchSnapshot();
    });
  });

  describe('REGISTER_SUCCESS', () => {
    it('should have no error and no pending state', () => {
      const createAction = new RegisterSuccess(authResp);

      const expectedResult = {
        error: null,
        pending: false,
      };

      const result = reducer(fromRegisterPage.initialState, createAction);

      expect(result).toMatchSnapshot();
    });
  });

  describe('REGISTER_FAILURE', () => {
    it('should have an error and no pending state', () => {
      const error = 'register failed';
      const createAction = new RegisterFailure(error);

      const expectedResult = {
        error: error,
        pending: false,
      };

      const result = reducer(fromRegisterPage.initialState, createAction);

      expect(result).toMatchSnapshot();
    });
  });
});
