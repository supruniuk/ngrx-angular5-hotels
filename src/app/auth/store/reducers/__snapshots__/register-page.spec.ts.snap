// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterPageReducer REGISTER should make pending to true 1`] = `
Object {
  "error": null,
  "pending": true,
}
`;

exports[`RegisterPageReducer REGISTER_FAILURE should have an error and no pending state 1`] = `
Object {
  "error": "register failed",
  "pending": false,
}
`;

exports[`RegisterPageReducer REGISTER_SUCCESS should have no error and no pending state 1`] = `
Object {
  "error": null,
  "pending": false,
}
`;

exports[`RegisterPageReducer undefined action should return the default state 1`] = `
Object {
  "error": null,
  "pending": false,
}
`;
